delta/plugins/archive/src/main/scala/ch/epfl/bluebrain/nexus/delta/plugins/archive/ArchiveDownload.scala

```diff
@@ -0,0 +1,189 @@
+          }.sequence
+        }
+      }
+
+      listIO.map { optEntries =>
+        val entries = optEntries.collect { case Some(value) => value } // discard None values
+        Source(entries).via(Archive.tar())
+      }
+    }
+
+    private def fetchFile(ref: FileReference, project: ProjectRef, ignoreNotFound: Boolean)(implicit
+        caller: Caller
+    ): IO[ArchiveRejection, Option[(TarArchiveMetadata, AkkaSource)]] = {
+      val refProject     = ref.project.getOrElse(project)
+      // the required permissions are checked for each file content fetch
+      val fileResponseIO = ref.ref match {
+        case ResourceRef.Latest(iri)           => files.fetchContent(IriSegment(iri), refProject)
+        case ResourceRef.Revision(_, iri, rev) => files.fetchContentAt(IriSegment(iri), refProject, rev)
+        case ResourceRef.Tag(_, iri, tag)      => files.fetchContentBy(IriSegment(iri), refProject, tag)
+      }
+      val tarEntryIO     = fileResponseIO
+        .map { fileResponse =>
+          val path     = pathOf(ref, project, fileResponse.filename)
+          val metadata = TarArchiveMetadata.create(path, fileResponse.bytes)
+          Some((metadata, fileResponse.content))
+        }
+        .mapError {
+          case _: FileRejection.FileNotFound     => ResourceNotFound(ref.ref, project)
+          case _: FileRejection.TagNotFound      => ResourceNotFound(ref.ref, project)
+          case _: FileRejection.RevisionNotFound => ResourceNotFound(ref.ref, project)
+          case FileRejection.AuthorizationFailed => AuthorizationFailed(ref.ref, project)
+          case other                             => WrappedFileRejection(other)
+        }
+      if (ignoreNotFound) tarEntryIO.onErrorRecover { case _: ResourceNotFound => None }
+      else tarEntryIO
+    }
+
+    private def fetchResource(ref: ResourceReference, project: ProjectRef, ignoreNotFound: Boolean)(implicit
+        caller: Caller,
+        col: AclCollection
+    ): IO[ArchiveRejection, Option[(TarArchiveMetadata, Source[ByteString, NotUsed])]] = {
+      val refProject = ref.project.getOrElse(project)
+      if (hasPermission(resources.read, refProject)) {
+        val tarEntryIO = resourceRefToByteString(ref, project).map { content =>
+          val path     = pathOf(ref, project)
+          val metadata = TarArchiveMetadata.create(path, content.length.toLong)
+          Some((metadata, Source.single(content)))
+        }
+        if (ignoreNotFound) tarEntryIO.onErrorHandle { _: ResourceNotFound => None }
+        else tarEntryIO
+      } else IO.raiseError(AuthorizationFailed(ref.ref, refProject))
+    }
+
+    private def resourceRefToByteString(
+        ref: ResourceReference,
+        project: ProjectRef
+    ): IO[ResourceNotFound, ByteString] = {
+      val p = ref.project.getOrElse(project)
+      val r = ref.representation.getOrElse(CompactedJsonLd)
+      Stream
+        .fromIterator[Task](exchanges.iterator)
+        .evalMap { _(p, ref.ref) }
```

Comment:
evalMap + collectFirst = `evalMapFilterValue`

Reply:
that works only for a ProjectionStream... it makes use of Message[A]; it doesn't work here

Reply:
I mean `evalMapFilter`